Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationWindow.cs

```diff
@@ -0,0 +1,417 @@
+                    {                        
+                        MigrateAllPrefabs();
+                        MigrateAllScenes();
+                    }
+
+                    using (new EditorGUILayout.HorizontalScope())
+                    {
+                        GUILayout.FlexibleSpace();
+
+                        string tooltip = "All Scenes and Prefabs Selected";
+                        EditorGUILayout.LabelField(new GUIContent(tooltip, InspectorUIUtility.WarningIcon));
+                    }
+                }
+            }
+        }
+
+        private void DrawMigrationObjectsList()
+        {
+            scrollPosition = EditorGUILayout.BeginScrollView(scrollPosition);
+
+            foreach (var migrationObject in migrationObjects)
+            {
+                using (var horizontal = new GUILayout.HorizontalScope())
+                {
+                    using (new EditorGUI.DisabledGroupScope(true))
+                    {
+                        EditorGUILayout.ObjectField(migrationObject, typeof(Object), false);
+                    }
+                    var icon = EditorGUIUtility.IconContent("winbtn_win_min_h");
+
+                    if (GUILayout.Button(icon, GUILayout.Width(30)))
+                    {
+                        migrationObjects.Remove(migrationObject);
+                        break;
+                    }
+                }
+            }
+            EditorGUILayout.Space();
+
+            using (new GUILayout.VerticalScope(EditorStyles.helpBox))
+            {
+                if (migrationObjects.Count == 0)
+                {
+                    GUI.enabled = false;
+                }
+
+                if (GUILayout.Button("Migrate"))
+                {
+                    MigrateSelection();
+                }
+
+                using (new EditorGUILayout.HorizontalScope())
+                {
+                    if (migrationObjects.Count > 0)
+                    {
+                        GUILayout.FlexibleSpace();
+
+                        string tooltip = $"{migrationObjects.Count} {toolbarTitles[(int)selectedToolbar]} selected for migration";
+                        EditorGUILayout.LabelField(new GUIContent(tooltip, InspectorUIUtility.WarningIcon));
+                    }
+                }
+            }
+            EditorGUILayout.EndScrollView();
+        }
+
+        private void AddSelection()
+        {
+            if (selectedToolbar == ToolbarOption.Objects)
+            {
+                if (!migrationObjects.Contains(selection))
+                {
+                    migrationObjects.Add(selection);
+                }
+            }
+            else
+            {
+                if (!migrationObjects.Contains(selection) && IsExtension(AssetDatabase.GetAssetPath(selection), SceneExtension))
+                {
+                    migrationObjects.Add(selection);
+                }
+            }
+        }
+
+        private void MigrateAllScenes()
```

Comment:
I think there may be value in separating the migration logic here from the gui logic and put it in its own class. A class of static migration helper functions may be useful. Not only would it be easier to write a test for an asset migration, but it would be possible in future to create command line functionality around migration that people could use in automatic migration scripts.

I'm thinking along similar lines to the MRTK Build Window, which has separated build logic from the GUI so that we can do command line builds.

Reply:
To be clear, I'm thinking of the following functions:
- MigrateAllScenes
- MigrateScene
- MigrateAllPrefabs
- MigrateSelection (repurposed and renamed as MigrateListOfObjects or something)
- MigrateGameObjectHierarchy

Reply:
Yeah... This is prob the only way to test and use the migration tool as a button on custom editors of obsolete components